fdbserver/ApplyMetadataMutation.h

```diff
@@ -22,22 +22,27 @@
 #define FDBSERVER_APPLYMETADATAMUTATION_H
 #pragma once
 
-#include "fdbclient/MutationList.h"
-#include "fdbclient/SystemData.h"
 #include "fdbclient/BackupAgent.actor.h"
+#include "fdbclient/MutationList.h"
 #include "fdbclient/Notified.h"
+#include "fdbclient/SystemData.h"
 #include "fdbserver/IKeyValueStore.h"
-#include "fdbserver/LogSystem.h"
 #include "fdbserver/LogProtocolMessage.h"
+#include "fdbserver/LogSystem.h"
 #include "fdbserver/ProxyCommitData.actor.h"
 
 inline bool isMetadataMutation(MutationRef const& m) {
 	// FIXME: This is conservative - not everything in system keyspace is necessarily processed by
 	// applyMetadataMutations
-	return (m.type == MutationRef::SetValue && m.param1.size() && m.param1[0] == systemKeys.begin[0] &&
-	        !m.param1.startsWith(nonMetadataSystemKeys.begin)) ||
-	       (m.type == MutationRef::ClearRange && m.param2.size() > 1 && m.param2[0] == systemKeys.begin[0] &&
-	        !nonMetadataSystemKeys.contains(KeyRangeRef(m.param1, m.param2)));
+	if (m.type == MutationRef::SetValue) {
+		return m.param1.size() && m.param1[0] == systemKeys.begin[0] &&
+		       !m.param1.startsWith(nonMetadataSystemKeys.begin);
+	} else if (m.type == MutationRef::ClearRange) {
+		return m.param2.size() > 1 && m.param2[0] == systemKeys.begin[0] &&
+		       !nonMetadataSystemKeys.contains(KeyRangeRef(m.param1, m.param2));
+	} else {
+		return false;
+	}
```

Comment:
Nit: This is more readable, but I am not sure if the original code is more efficient -- avoiding `if-else` conditions. Maybe compiler is smart enough to optimize to the same speed.

Reply:
The interesting part is, `clang` will generate one extra `jmp` for the first `if` branch for the following code.

```
struct MutationRef {
	int type;
	char param1[10] = "abcdefghi";
	char param2[10] = "jklmnopqr";
};


bool test0(MutationRef m) {
	if (m.type == 1) {
		return strlen(m.param1) && m.param1[0] == 'a';
	} else if (m.type == 2) {
		return strlen(m.param2) > 1 && m.param2[0] == 'b';
	} else {
		return false;
	}
}

bool test1(MutationRef m) {
	return (m.type == 1 && strlen(m.param1) && m.param1[0] == 'a') || (m.type == 2 && strlen(m.param2) > 1 && m.param2[0] == 'b');
}
```

Generated assembly:

```
__Z5test011MutationRef:                 ## @_Z5test011MutationRef
	.cfi_startproc
## %bb.0:
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset rbp, -16
	mov	rbp, rsp
	.cfi_def_cfa_register rbp
	push	rbx
	push	rax
	.cfi_offset rbx, -24
	lea	rbx, [rbp + 16]
	mov	eax, dword ptr [rbp + 16]
	cmp	eax, 2
	je	LBB0_3
## %bb.1:
	cmp	eax, 1
	jne	LBB0_5
## %bb.2:
	lea	rdi, [rbx + 4]
	call	_strlen
	test	rax, rax
	setne	cl
	cmp	byte ptr [rbx + 4], 97
	jmp	LBB0_4
LBB0_3:
	lea	rdi, [rbx + 14]
	call	_strlen
	cmp	rax, 1
	seta	cl
	cmp	byte ptr [rbx + 14], 98
LBB0_4:
	sete	al
	and	al, cl
                                        ## kill: def $al killed $al killed $eax
	add	rsp, 8
	pop	rbx
	pop	rbp
	ret
LBB0_5:
	xor	eax, eax
                                        ## kill: def $al killed $al killed $eax
	add	rsp, 8
	pop	rbx
	pop	rbp
	ret
	.cfi_endproc
                                        ## -- End function
	.globl	__Z5test111MutationRef          ## -- Begin function _Z5test111MutationRef
	.p2align	4, 0x90
__Z5test111MutationRef:                 ## @_Z5test111MutationRef
	.cfi_startproc
## %bb.0:
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset rbp, -16
	mov	rbp, rsp
	.cfi_def_cfa_register rbp
	push	rbx
	push	rax
	.cfi_offset rbx, -24
	lea	rbx, [rbp + 16]
	mov	eax, dword ptr [rbp + 16]
	cmp	eax, 2
	je	LBB1_3
## %bb.1:
	cmp	eax, 1
	jne	LBB1_6
## %bb.2:
	lea	rdi, [rbx + 4]
	call	_strlen
	test	rax, rax
	setne	cl
	cmp	byte ptr [rbx + 4], 97
	sete	al
	and	al, cl
	add	rsp, 8
	pop	rbx
	pop	rbp
	ret
LBB1_3:
	add	rbx, 14
	mov	rdi, rbx
	call	_strlen
	cmp	rax, 2
	jb	LBB1_6
## %bb.4:
	cmp	byte ptr [rbx], 98
	sete	al
                                        ## kill: def $al killed $al killed $eax
	add	rsp, 8
	pop	rbx
	pop	rbp
	ret
LBB1_6:
	xor	eax, eax
                                        ## kill: def $al killed $al killed $eax
	add	rsp, 8
	pop	rbx
	pop	rbp
	ret
	.cfi_endproc
                                        ## -- End function
```

One extra `jmp` for the `if` implementation. I would guess it is not a huge performance impact?

Reply:
It's probably fine. With deep optimization levels, the difference may not exist. As discussed in yesterday's meeting, the key is to preserve the evaluation order, which you did. So the change looks good to me.

Reply:
If the evaluation order is crucial here, can we add a giant warning comment to remind anyone touching the file/function?

Reply:
This code is actually a bit confusing, since I thought operator && would shortcut, i.e. jump to false when the first operand is false. However, it actually evaluates *all* operands. The compiler knows the function calls have no side effect, which might be smart, or not. The code is already optimized using the `-O3` flag, perhaps there are some flags like '-ffast-math' which allows more aggressive optimizations, but I guess we could skip look up the manual at this stage.

Reply:
> If the evaluation order is crucial here, can we add a giant warning comment to remind anyone touching the file/function?

I believe the order is not crucial, as `size()`, `operator[]` and `startsWith` have no side effects.